docs/dev/design/better-doc-urls-handling.rst

```diff
@@ -0,0 +1,417 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Projects:
+
+- project (latest, 1.0)
+- project-es (latest, 1.0)
+- subproject (latest)
+
+.. list-table:: 
+   :header-rows: 1
+
+   * - Request
+     - Requested file
+     - Current project
+     - Note
+   * - /projects/subproject/manual/index.html
+     - /latest/manual/index.html
+     - subproject
+     -
+   * - /projects/subproject/en/latest/manual/index.html
+     - 404
+     - subproject
+     - The subproject is single version
+   * - /projects/subproject/404
+     - 404
+     - subproject
+     - The subproject exists, but the file doesn't
+   * - /projects/subproject/br/latest/manual/index.html
+     - /latest/br/latest/manual/index.html
+     - subproject
+     - The subproject has a ``br`` directory!
+   * - /projects/nothing/manual/index.html
+     - 404
+     - project
+     - The subproject doesn't exist
+   * - /en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     -
+   * - /404
+     - 404
+     - project
+     -
+
+Project with custom prefix
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- project (latest, 1.0)
+- subproject (latest, 1.0)
+
+``project`` has the ``prefix`` prefix, and ``sub`` subproject prefix.
+
+.. list-table:: 
+   :header-rows: 1
+
+   * - Request
+     - Requested file
+     - Current project
+     - Note
+   * - /en/latest/manual/index.html
+     - 404
+     - project
+     - The prefix doesn't match
+   * - /prefix/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     -
+   * - /projects/subproject/en/latest/manual/index.html
+     - 404
+     - project
+     - The subproject prefix doesn't match
+   * - /sub/subproject/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - subproject
+     -
+   * - /sub/nothing/en/latest/manual/index.html
+     - 404
+     - project
+     - The subproject doesn't exist
+
+Project with custom subproject prefix (empty)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- project (latest, 1.0)
+- subproject (latest, 1.0)
+
+``project`` has the ``/`` subproject prefix,
+this allow us to serve subprojects without using a prefix.
+
+.. list-table:: 
+   :header-rows: 1
+
+   * - Request
+     - Requested file
+     - Current project
+     - Note
+   * - /en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     -
+   * - /projects/subproject/en/latest/manual/index.html
+     - 404
+     - project
+     - The subproject prefix doesn't match
+   * - /subproject/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - subproject
+     -
+   * - /nothing/en/latest/manual/index.html
+     - /latest/manual/index.html
+     - project
+     - The subproject/file doesn't exist
+
+Questions
+---------
+
+- When using custom URLs,
+  should we support changing the URLs
+  that aren't related to doc serving?
+
+  These are:
+
+  - Health check
+  - Proxied APIs
+  - robots and sitemap
+  - The ``page`` redirect
```

Comment:
Maybe... The main use case here would be prefixing I think, as well as users proxying to us. We have hit issues in the past where we assume that we own the entire domain's hosting (`docs.example.org/_/`), but when there's a prefix, we likely want to serve it at `{prefix}/_/`, in case we aren't hosting the root domain in some configuration.

Reply:
Ohh, yes, then people that proxy us are the main target of the custom prefix feature I'd say, and others will probably use it to change the subprojects prefix and use the normal serving.